Improve shadow filters

x2048 opened the pull request:
With the introduction of new SM texture distortion in #12166, PCF filter for shadows was outdated and produced wrong results. 

What's done:
* Use information available in the new vertex shaders from #12191 (perspective_factor), instead of trying to reverse-calculate it from the final position.
* Use fewer filter samples, as it is no longer necessary with the new, more accurate distortion, the same applies to penumbra calculation.
* Optimize the filter convolution loops to perform few repetitive calculations.

Benefits:
* Filtering and less 'dancing' of shadows when texture precision permits that.
* Soft shadows work again.
* It's faster than the previous implementation.

## To do

This PR is a Work in Progress and depends on #12191 being merged, but review is welcome.

- [x] Rebase when #12191 is merged.
- [x] Fix artifacts for certain combinations of overlapping distant and close shadows.
        - So far it has been solved by limiting the soft shadow radius

## How to test

1. Enable dynamic shadows at your favorite quality level.
2. Optional: You can also set the soft shadow radius to something more than 1.
3. Start a compatible game
4. Enjoy the filtered/soft shadows

x2048 committed:
Rewrite shadow filtering for the new distortion

lhofhansl commented:
Is there any way to make the shadow boundaries softer?

This is with view distance 300 and shadow texture size 1024:
![screenshot_20220414_113624](https://user-images.githubusercontent.com/446130/163455035-c6d03c40-9a39-48a5-a173-72e5fe886537.png)

Would be nice if that can look smoother.

lhofhansl commented on client/shaders/nodes_shader/opengl_fragment.glsl:
```diff
@@ -116,23 +117,17 @@ float getHardShadow(sampler2D shadowsampler, vec2 smTexCoord, float realDistance
 #if SHADOW_FILTER == 2
-	#define PCFBOUND 3.5
-	#define PCFSAMPLES 64.0
+	#define PCFBOUND 1.5 // 4x4
```
Do we need to make the same changes for the object shader?

x2048 replied:
I want to make all the improvements in the nodes shader first, then transplanting the changes to objects shader is a single commit.

lhofhansl commented:
Looks like the moire pattern are worse now:

![screenshot_20220414_154953](https://user-images.githubusercontent.com/446130/163490091-dde20867-4a2e-408c-8c35-83031f6bac23.png)

This is the Hidroplane mode. (https://content.minetest.net/packages/apercy/hidroplane/)

x2048 committed:
Change filter radius calulation

* Avoid peter-panning effect due to filtering of short shadows
* Scale the penumbra radius linearly
* Try a different penumbra detection algorithm

x2048 commented:
The moire pattern on objects might get worse due to changes in offset biases that need more tuning (I spotted some artifacts too). Latest changes to combat peter-panning might also contribute to more visible moire/self-shadowing.

lhofhansl commented:
> Fix artifacts for certain combinations of overlapping distant and close shadows.

I noticed that with soft-shadows. When a close shadows overlaps a distance soft-shadow the far (softer) shadow "wins". Is that what you mean?

lhofhansl commented:
Filtering with latest change:

![screenshot_20220420_172848](https://user-images.githubusercontent.com/446130/164345816-1189a4a2-ebd3-44f1-bcfa-fca6a2c12ae4.png)

x2048 committed:
More adjustments for soft shadows

lhofhansl commented on client/shaders/nodes_shader/opengl_fragment.glsl:
```diff
@@ -153,44 +153,45 @@ float getPenumbraRadius(sampler2D shadowsampler, vec2 smTexCoord, float realDist
 
-	if (depth == 10.0)
-		return 0.0;
+	depth *= 1e0 * world_to_texture / max(n, 1.0);
```
ERROR[Main]: Irrlicht: 0(239) : error C1101: ambiguous overloaded function reference "max(int, float)"

x2048 replied:
Fixed in the latest commits

x2048 committed:
Scale filter radius by a fixed value.

Reduces the amount of shadow aliasing noise

lhofhansl commented on client/shaders/nodes_shader/opengl_fragment.glsl:
```diff
@@ -188,10 +189,9 @@ float getPenumbraRadius(sampler2D shadowsampler, vec2 smTexCoord, float realDist
 
-	depth *= 1e0 * world_to_texture / max(n, 1.0);
-
+	depth *= world_to_texture / max(n, 1);
```
Getting this now:
ERROR[Main]: Irrlicht: 0(241) : error C1101: ambiguous overloaded function reference "max(int, int)"

Don't you just love shaders? :(

lhofhansl replied:
```
ERROR[Main]: Irrlicht: 0(241) : error C1101: ambiguous overloaded function reference "max(int, int)"
(0) : gp5 cp50 fp50 vp50 gp50 cpf fpf vpf gpf superp float max(superp float, superp float)
(0) : gp5 mediump float max(mediump float, int)
(0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf lowp float max(lowp float, lowp float)
(0) : ps gp5 gp4 cp50 fp50 vp50 gp50 cpf fpf vpf gpf mediump float max(mediump float, mediump float)
(0) : float max(float, float)
```

x2048 replied:
Fixed in the latest commits.

lhofhansl commented:
Peter panning almost gone:
![screenshot_20220423_171151](https://user-images.githubusercontent.com/446130/164950272-3c296a25-f052-41c9-a207-ba63d475fe57.png)

Traded for no close filtering I guess:
![screenshot_20220423_171202](https://user-images.githubusercontent.com/446130/164950275-932dce10-44b0-441e-a98c-cd054d8399d2.png)

x2048 committed:
Add adaptive filter quality for soft shadows

x2048 commented:
>>   Fix artifacts for certain combinations of overlapping distant and close shadows.

> I noticed that with soft-shadows. When a close shadows overlaps a distance soft-shadow the far (softer) shadow "wins". Is that what you mean?

There was a more specific problem that I would call a "tree on a cliff" problem. When a high tree projected a shadow on a cliff, its silhouette would appear in the shadow of the cliff, because those pixels had much longer shadow depth and penumbra radius. The problem should be fixed now by translating depth to penumbra using a linear function (opposed to quadratic function previously).

The problem of softer shadows taking over the harder one is due to storing only one depth value per texel, and I don't see how it can be solved without complicating the SM texture generation significantly.

x2048 commented:
> Peter panning almost gone:
> Traded for no close filtering I guess:

Yes, that's the tradeoff I chose to do for near-zero shadow depth.

x2048 commented:
> Is there any way to make the shadow boundaries softer?

Should be done in the latest commits. Filter radius is now at least 2 texels, which together with adaptive quality reduces effect of the shard edges / dancing shadows.

x2048 committed:
Align filters in nodes and objects shaders

lhofhansl commented:
So far I think it still looks better without this change. :(

x2048 commented:
Can you share examples?

x2048 committed:
Tune sharpness_factor to avoid unfiltered shadows

x2048 commented:
@lhofhansl I've tried to find the balance between player's feet and peter panning. Looks much better on my machine at least.

lhofhansl commented:
Thanks for looking!

The fragment shaders do not compile for me. I had to make the following changes:
-	int n = 0;
+	float n = 0;
...
-	depth *= world_to_texture / max(n, 1);
+	depth *= world_to_texture / max(n, 1.0);

and
-	samples = clamp(int(pow(2 * radius + 1, 2.0)), 1, samples);
+	samples = int(clamp(int(pow(2 * radius + 1, 2.0)), 1, samples));

At the relevant places on the object and nodes fragment shaders.

Also the player shadow looks still "jaggedy":

![screenshot_20220501_131042](https://user-images.githubusercontent.com/446130/166162898-94bba14e-7f59-4311-bc32-f3e505b5102d.png)

x2048 committed:
Fix compatibility errors in fragment shaders

x2048 commented:
@lhofhansl  I've applied the fixes, thank you.

> Also the player shadow looks still "jaggedy":

I don't get the same effect as you do.
![image](https://user-images.githubusercontent.com/4933697/166335758-982c6c90-8cfa-4b8a-9369-de022e82ba0d.png)

lhofhansl commented:
Hmm... I think my view distance is larger (300) and shadow map size is set 4096 (and max shadow distance is 450)

lhofhansl commented:
With the newest:
![screenshot_20220505_171342](https://user-images.githubusercontent.com/446130/167046553-9878c6bb-217b-49b7-9ea5-ae8ad5f9cae4.png)

x2048 committed:
Reduce white halo around sharp shadows

x2048 commented:
I've adjusted the math to reduce the thin white halo and unnecessary blurring when the shadow is sharp.

lhofhansl commented:
Thanks for sticking with this @x2048. I'll try out the latest change.

lhofhansl commented:
Closeby shadows look very pixelate still. (shadow map size 2048, viewing_range 300, max shadow distance 450).

![screenshot_20220507_153959](https://user-images.githubusercontent.com/446130/167274087-ad091b2e-a3fd-499f-93a2-b7dbb9b3ab75.png)

I always think that _more_ filtering looks better and immerses better, since you notice the artifacts less.

lhofhansl commented:
Don't give up on us @x2048  :)

x2048 commented:
Still here. Keeping silence since there's not a lot I can share now

x2048 committed:
Small improvements to filters

Adjust sharpness factor for perspective factor
Avoid sharp shadows on surfaces without normals (e.g. plants)

x2048 committed:
Remove optimization for small sharpness_factor

x2048 committed:
Disable shadow debug output

x2048 committed:
Disable all filtering on SM texture to avoid artifacts

x2048 committed:
Calculate penumbra radius using a single sample

x2048 commented:
@lhofhansl OK, I have something interesting now.
1. Fixed a problem with OpenGL filtering screwing depth values and leading to artifacts at shadow edges.
2. Changed penumbra radius calculation to O(1) so now soft shadow radius can be adjusted without performance penalty.
3. Edgy feet are still here but not that edgy. I prefer to avoid peter panning more than proper shadow for own feet.

lhofhansl commented:
:+1:

With soft-shadows set to 10 is looks nice now! (Still the very close edges with the player shadows, but I think that is cool)
Might want to default soft-shadows to 5 or something, and also allow larger values.

x2048 committed:
Increase default and maximum soft shadow radius

x2048 committed:
Copy latest fixes to object shaders

x2048 commented:
Agree. I've increased the maximum blur to 30 now, changed default to 5 and copied the the changes to object shaders.

x2048 commented:
Some screenshots:
* Duck plane:
  ![image](https://user-images.githubusercontent.com/4933697/168494132-ca5e6449-7d08-4e57-ba6f-9aa4e39178fe.png)
* Colored windows
  ![image](https://user-images.githubusercontent.com/4933697/168494343-ca3cd19a-08d9-4457-9558-a9c54f1465e9.png)

kilbith commented:
@x2048 Can you also change the default value of `shadow_update_frames` to `1` because I really can stand this "laggy" feel with shadows when the map gets changed. It really does look bad.

x2048 committed:
Add line directive to all shaders

Helps troubleshooting

x2048 committed:
Fix default and bounds for soft shadow radius

Default is 5.0
Maximum is 15.0 to avoid phantom shadows

lhofhansl commented on builtin/settingtypes.txt:
```diff
@@ -631,8 +631,8 @@ shadow_update_frames (Map shadows update frames) int 8 1 16
-#    Minimum value: 1.0; maximum value: 30.0
-shadow_soft_radius (Soft shadow radius) float 1.0 5.0 30.0
+#    Minimum value: 1.0; maximum value: 15.0
+shadow_soft_radius (Soft shadow radius) float 1.0 5.0 15.0
```
You mean `shadow_soft_radius (Soft shadow radius) float 5.0 1.0 15.0`?
(it's default min max)

x2048 replied:
Done

sfan5 commented on client/shaders/3d_interlaced_merge/opengl_fragment.glsl:
```diff
@@ -1,3 +1,5 @@
+#line 1
```
You know that shader.cpp could automatically prepend this, right?

x2048 replied:
Kind of realized afterwards :)

x2048 replied:
It's the proper way now.

x2048 committed:
Revert "Add line directive to all shaders"

This reverts commit 198a6f8537fd225f859ba28fc5f6aa315e2c90c6.

x2048 committed:
Make line numbers in shader errors match the code

x2048 committed:
Fix default and minimum for soft_shadow_radius

Zughy commented:
@x2048 is this still WIP?

x2048 commented:
@Zughy Thanks for pointing out, it's approved now, so not a WIP
